src/rules/cognitive-complexity.ts

```diff
@@ -339,20 +339,47 @@ const rule: TSESLint.RuleModule<string, (number | 'metric' | 'sonar-runtime')[]>
     }
 
     function visitLogicalExpression(logicalExpression: TSESTree.LogicalExpression) {
+      if (isJsxShortCircuitNode(logicalExpression)) {
+        return;
+      }
+
       if (!consideredLogicalExpressions.has(logicalExpression)) {
         const flattenedLogicalExpressions = flattenLogicalExpression(logicalExpression);
 
         let previous: TSESTree.LogicalExpression | undefined;
         for (const current of flattenedLogicalExpressions) {
           if (!previous || previous.operator !== current.operator) {
-            const operatorTokenLoc = getFirstTokenAfter(logicalExpression.left, context)!.loc;
+            const operatorTokenLoc = getFirstTokenAfter(current.left, context)!.loc;
             addComplexity(operatorTokenLoc);
           }
           previous = current;
         }
       }
     }
 
+    function isJsxShortCircuitNode(logicalExpression: TSESTree.LogicalExpression) {
+      const isShortCircuit =
+        logicalExpression.parent?.type === 'JSXExpressionContainer' &&
+        isJsxShortCircuitSubNode(logicalExpression, logicalExpression.left) &&
+        isJsxShortCircuitSubNode(logicalExpression, logicalExpression.right);
+      if (isShortCircuit) {
+        consideredLogicalExpressions.add(logicalExpression);
+      }
+      return isShortCircuit;
+    }
+
+    function isJsxShortCircuitSubNode(root: TSESTree.LogicalExpression, node: TSESTree.Node) {
```

Comment:
So the maximum of allowed binary operators is 2?

Reply:
As discussed, the last commit adds recursion logic to the short-circuit check.